### ISSUE TITLE
Max Callstack and Unsubscribe Issues with rxjs > beta 12

### ISSUE THREAD

**Issue by MadMub** [ReactiveX/rxjs#2146]
**RxJS version:**
greater than beta 12 (all release candidates)
**Code to reproduce:**
https://github.com/MadMub/rxjs-max-callstack
**Expected behavior:**
nodejs callstack should not blow up, all client connections close properly
**Actual behavior:**
the callstack is overloaded, with smaller numbers of connections not all events seem to fire? (see demo). In beta 12 I'm able to connect and drop over roughly 10x the connections.
**Additional information:**
The README on the repo provides context

**Comment by blesh** [comment]
If I had to guess, I'd say `client.destroy()` synchronously triggers another event, either `error` or `close`. I haven't run the demo yet, but that's my first inclination.

Also, if I were to implement something like this, I think I'd go more the route of wrapping the client in a single observable.

```js
new Observable(observer => {
  const nextHandler = d => observer.next(d);
  const errorHandler = e => observer.error(e);
  const completeHandler = hasError => hasError ? observer.error(hasError) : observer.complete();

  client.addEventListener('data', nextHandler);
  client.addEventListener('error', errorHandler);
  client.addEventListener('close', completeHandler);

  return () => {
    client.removeEventListener('data', nextHandler);
    client.removeEventListener('error', errorHandler);
    client.removeEventListener('close', completeHandler);
    if (needDestroyed) { // TODO: logic for this decision here?
      client.destroy();
    }
  };
});
```

Just as a general reference. It will be more verbose, but it might be more correct for you, and should definitely perform better.

**Comment by MadMub** [comment]
@blesh thanks for single observable idea, in practice I think that is a way better idea and I will switch to that. I'm still concerned by what's happening though so I'd like to ask a few questions:

1) you mentioned something about destroy potentially synchronously making `close` emit, if that were the case, is it bad that i don't see this issue on beta 12, should I be able to hit max callstack somewhat easily on > rc1? Is there special condiseration with events firing too close together?

2) I will need to confirm it, but I think when I close the dummy clients script the server isn't firing error ever, simply `close`. How come rxjs beta 12 can support unsubscribing 4k connections but > rc1 only a few hundred. My next step is to try and look into this issue myself , step through the commits made between beta 12 and rc1.

I love rxjs and its hard for me not to use it for event patterns (i'm hooked!) but my current use case does require pretty high performance. Thank you for the feedback hopefully I can contribute and fix this problem (or identify if its something erroneous on my end and usage of rxjs)

**Comment by blesh** [comment]
@MadMub if performance is a critical concern, I'd definitely look into my recommendation above. 

I'll try to get a chance to look at this issue more in-depth soon.

**Comment by MadMub** [comment]
ping @blesh , months later I upgraded to rxjs5.2  and started to bring rxjs back into my backend server. The server is a socket proxy server and I use an observable on an event emitter to perform logic when "timeout" is fired (a socket has no bytes received in a while).

When connecting 50 sockets and letting them sit idle for 30 seconds (the configured timeout), all 50 timeouts fire, however only 3-7 of the callbacks/observable chains rigged up to those events actually fire, the other ~40 leak. When I roll back to beta.12 (which at this point is quite out of date) the issue goes away. At an even bigger scale (more than 400 sockets), I think the leak leads to an unexpected crash (error message is Unsubscribe error, maxcallstack something...)

As I mentioned in my initial description something changed between beta.12 and rc1 around core subscribe/unsubscribe? At this point if I want to rely more on rxjs I need to try and hunt this down myself. I just figured I'd chime in and inform that the issue is still present.

**Comment by trxcllnt** [comment]
@MadMub this issue has been fixed by merging #2238, but hasn't been published to npm yet. If you change the `subscribe()` calls in your demo repo to be `subscribe({})`, it should work (at least, it did for me).

**Comment by MadMub** [comment]
Okay, I will wait till next release. Glad this one has been solved, wish I had chased it down more back in November thanks for the update!

**Comment by kwonoj** [comment]
I think this issue can be closed now via #2238 and other suggestions. Feel freely reopen if problem persists.
